### PR TITLE
gh-89480: Document the gdb helpers

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-89480

Content adapted from https://devguide.python.org/development-tools/gdb/# and https://wiki.python.org/moin/DebuggingWithGdb

----
📚 Documentation preview 📚: https://cpython-previews--115657.org.readthedocs.build/

### CODE REVIEW

**Comment by encukou** [comment]
Thank you! I'll update this for recent GDB, and then we can perhaps point the devguide to it.

To everyone else: Please don't add GitHub suggestions, they might be lost.

**Comment by encukou** [comment]
I'm signing of for the day. Here's a first pass, I'll finish tomorrow.
It's safe to add suggestions now, if anyone wants to do that.

**Comment by encukou** [comment]
Ready for review.

Diátaxis-wise, this leans toward being a *reference* rather than a *how-to guide*. I'm not sure how to quickly handle that -- the document assumes you know debugging with GDB, and only provides a few details.
IMO, adding it would be an improvement, and restructuring can be done later.

Once merged, the [devguide page](https://devguide.python.org/development-tools/gdb/) should be updated to point here, and retain only the parts specific to CPython core (`test_gdb` & tips).

File: Doc/howto/gdb_helpers.rst
Comment (hugovk) [inline Doc/howto/gdb_helpers.rst]:
```suggestion
   If the current C frame corresponds to multiple Python frames, ``py-print``
```

File: Doc/howto/gdb_helpers.rst
Comment (hugovk) [inline Doc/howto/gdb_helpers.rst]:
```suggestion
The extension commands complement GDB's built-in commands.
```

File: Doc/howto/gdb_helpers.rst
Comment (hugovk) [inline Doc/howto/gdb_helpers.rst]:
```suggestion
   If the current C frame corresponds to multiple Python frames, locals from
```

**Review by hugovk — COMMENTED** [review]
Some typos found with https://github.com/codespell-project/codespell

File: Doc/howto/gdb_helpers.rst
Comment (encukou) [reply Doc/howto/gdb_helpers.rst]: Arrrr!

Thank you :)

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst]: By looking at the ToC, it's not clear to me what these two ways are.

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst]: This title is not very descriptive -- maybe
```suggestion
Installing debug information package on Linux
```
If this only works on Linux, you could just use:
```suggestion
Installing Python debug information package
```
and state in the first paragraph that these are only available on Linux.

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst]: If this is a prerequisite, it should be listed above, and this section would just explain how to do it, with a title like:
```suggestion
Building Python from source
```

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst]:
```suggestion
   If you do not see instructions for your version of GDB, put this in your
```

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst]: This title could also be more "action-oriented":
```suggestion
Building Python in "debug" and "development" mode
```
or
```suggestion
Enabling "debug" and "development" mode
```
This could also be a sub-section of the "Building Python from source" section above.  You wouldn't want to read that and build everything and then discover that you have to do it again with debug/dev mode enabled.

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst:114]: This seems to be the actual how-to.

You could also add a short introductory paragraph like "After installing the dependencies, you can now use the `python-gdb extension to ...".

File: Doc/howto/gdb_helpers.rst
Comment (ezio-melotti) [inline Doc/howto/gdb_helpers.rst]:
```suggestion
=====================
```
This broke the formatting of the title -- maybe this is the second way mentioned in the first paragraph?

**Review by ezio-melotti — COMMENTED** [review]
I just skimmed through it and left a few comments about the high-level structure of the document and its headers.

**Review by willingc — CHANGES_REQUESTED** [review]
Let's come up with a better title and address the introduction section that I mention in my comments. After that, I think this will be ready to merge from my vantage point.

I'm less concerned about where this falls diataxis-wise and more concerned that we have clarity of its purpose for the reader and for future disambiguation of the old-style HOWTOs.

File: Doc/howto/gdb_helpers.rst
Comment (encukou) [reply Doc/howto/gdb_helpers.rst]: You do want both: for debugging *extensions* it's sometimes easier to use the production build.
And these are not specific to source builds.

File: Doc/howto/gdb_helpers.rst
Comment (encukou) [reply Doc/howto/gdb_helpers.rst:114]: The previous section tells you how to set things up, this tessl you how to use it :)

File: Doc/howto/gdb_helpers.rst
Comment (encukou) [reply Doc/howto/gdb_helpers.rst]: Right, there's only one way left in the document. (The other was for the 2010's.)
Thanks for the catch!

File: Doc/howto/gdb_helpers.rst
Comment (encukou) [reply Doc/howto/gdb_helpers.rst]: Setting things up depends on how you got your Python. These are examples for two common cases: source builds, and Linux distros. (No Windows or Mac, you wouldn't typically use GDB there.)

File: Doc/howto/gdb_helpers.rst
Comment (encukou) [reply Doc/howto/gdb_helpers.rst]: I'd keep Python itself is a prerequisite for the prerequisites. The guide doesn't tell you how to get it, but how to set things up for it.
Hopefully “Setup with Python built from source” is clearer.

**Comment by encukou** [comment]
Thanks! Your perspectives are very helpful here; I'm too deep in :)
I want to move it forward, it doesn't really need to be quick. It's a deferred release blocker for 3.11, a few more months won't matter.

@willingc 
> guidance for the reader
>
>  -  who would find this content useful
>  -  when would they use gdb and specifically the python helper
>  -  what will they learn if they read the rest of the document

I thought the first section covers exactly these questions. When you have the time, could you please be more specific about how it should be written?

**Comment by willingc** [comment]
@encukou I like the changes that you have made to the headings.

> I thought the first section covers exactly these questions. When you have the time, could you please be more specific about how it should be written?

I think it is being more explicit to the reader. I will leave an example in the comments here later today. Thanks!

**Comment by willingc** [comment]
@encukou This is more along the lines of what I was thinking:
- high-level what
- why this is needed
- who would use this information
It takes most of the info that you wrote but breaks it into more logical chunks for a reader, such a scientist writing and debugging a C extension, who has not yet used the the Python extension with GDB.

```
This document explains how GDB and the Python extension, ``python-gdb.py``, can
be used together to debug CPython extensions and the CPython interpreter itself.

When debugging low-level problems such as crashes or deadlocks, a low-level debugger,
such as GDB, is useful to diagnose and correct the issue. By default, GDB (or any
of its front-ends) doesn't support high-level information specific to the CPython
interpreter.

The ``python-gdb.py`` extension adds CPython interpreter information to GDB. The
extension helps introspect a currently executing Python function. Given
a Python object represented by a standard PyObject * pointer, the extension surfaces
the type and value of the object.

Developers who are working on CPython extensions or tinkering with parts of CPython
that are written in C can use this document to learn how to use GDB with the
``python-gdb.py`` extension.

.. note::
    
   This document assumes that you are familiar with the basics of GDB and the
   CPython C API. It consolidates guidance from the CPython devguide
   and the Python wiki.
```

**Review by willingc — COMMENTED** [review]
See sample wording in my comment.

**Review by willingc — APPROVED** [review]

**Comment by encukou** [comment]
Thank you for the concrete wording! I hope it will help future me write better introductions on my own :)

I edited your suggestion for technical details, and to highlight that this doesn't teach how to use GDB itself -- the guide (and the extension) are mainly useful for people who already use GDB.

I take your GitHub approval as a sign that you don't necessarily want to review again, so, I'll merge on Monday if there are no objections.

**Comment by encukou** [comment]
Thank you!

**Pull request merged**
